relayer/src/main.rs

```diff
@@ -40,68 +42,79 @@ impl Synchronizer {
-                    // Block has been saved
+                    current_height += block_to_submit_len;
                 }
                 _ => {
                     // network error after retries
                     panic!("Off-chain relay panics after multiple attempts to save block");
                 }
             }
-
-            current_height += 1;
         }
     }
 
-    // Adjust height of the block to start submitting new fork, which might become a new main
-    async fn adjust_height_to_the_fork(&self, current_height: u64) -> u64 {
-        let mut amount_of_blocks_to_request = 25;
-
-        // If we inspected 10_000 bitcoin blocks and still cannot find
-        // the point where fork happened something is very wrong
-        // it means it happened 10_000 * 10 minutes = 69 days ago (relayer was down for 69 days?)
-        while amount_of_blocks_to_request < 10_000 {
-            amount_of_blocks_to_request *= 2;
+    async fn get_last_correct_block_height(&self) -> Result<u64, Box<dyn std::error::Error>> {
+        let last_block_header = self.near_client.get_last_block_header().await?;
+        let last_block_height = last_block_header.block_height;
 
-            let last_block_hashes_in_relay_contract = self
-                .near_client
-                .receive_last_n_blocks(amount_of_blocks_to_request, 0)
-                .await
-                .expect("read block header successfully");
-
-            // Starting to look for diverge point from previous block
-            let mut height = current_height - 1;
+        if self.get_bitcoin_block_hash_by_height(last_block_height)
+            == last_block_header.block_hash.to_string()
+        {
+            return Ok(last_block_height);
+        }
 
-            for _i in 0..amount_of_blocks_to_request {
-                let block_from_bitcoin_node =
-                    self.bitcoin_client.get_block_header_by_height(height);
+        let last_block_hashes_in_relay_contract = self
+            .near_client
+            .get_last_n_blocks_hashes(self.config.max_fork_len, 1)
+            .await
+            .expect("read block header successfully");
 
-                let hash = block_from_bitcoin_node.block_hash().to_string();
+        let last_block_hashes_count = last_block_hashes_in_relay_contract.len();
 
-                // We found that this is the first block in current bitcoin node state that we also have
-                // in our main chain in smart contract state.
-                // This is a diverge point. We will start submitting new fork from this point.
-                if last_block_hashes_in_relay_contract.contains(&hash) {
-                    return height;
-                }
+        let mut height: u64 = last_block_height - 1;
 
-                height -= 1;
+        for i in 0..last_block_hashes_count {
+            if last_block_hashes_in_relay_contract[last_block_hashes_count - i - 1]
+                == self.get_bitcoin_block_hash_by_height(height)
```

Comment:
At each iteration, the height changes, and we make requests with different parameters. Therefore, this cannot be moved outside the loop

Reply:
Right. As far as I understand we're trying to find the latest block before the fork here. With `max_fork_len` configured to 500 I am worried about the number of requests we will make here in case of a large fork. I don't know how critical this is so happy to keep it this way for now